server/core_benchmarks_test.go

```diff
@@ -427,157 +426,161 @@ func BenchmarkCoreFanIn(b *testing.B) {
-							// wait till publishers are ready
-							publishersReadyWg.Wait()
+			defer ncPub.Close()
+		}
 
-							// start the clock
-							b.ResetTimer()
-							// wait till termination cond reached
-							benchCompleteWg.Wait()
-							// stop the clock
-							b.StopTimer()
+		for i := range publishers {
+			go func(pubId int) {
+				publisher := publishers[pubId]
+				subject := fmt.Sprintf("%s.%d", subjectBaseName, pubId)
+
+				// signal that this publisher has been torn down
+				defer finishedPublishersWg.Done()
+
+				// wait till all other publishers are ready to start workload
+				publishersReadyWg.Wait()
+
+				// publish until quitCh is closed
+				for {
+					select {
+					case <-publisher.quitCh:
+						return
+					default:
+						// continue publishing
+					}
+					err := publisher.conn.Publish(subject, messageData)
+					if err != nil {
+						publisher.publishErrors += 1
+					}
+				}
+			}(i)
+		}
 
-							// send quit signal to all publishers
-							for i := range publishers {
-								publishers[i].quitCh <- true
-							}
-							// wait for all publishers to shutdown
-							finishedPublishersWg.Wait()
+		// set bytes per operation
+		b.SetBytes(messageSize)
+		// main thread is ready
+		publishersReadyWg.Done()
+		// wait till publishers are ready
+		publishersReadyWg.Wait()
+
+		// start the clock
+		b.ResetTimer()
+		// wait till termination cond reached
+		benchCompleteWg.Wait()
+		// stop the clock
+		b.StopTimer()
+
+		// send quit signal to all publishers
+		for i := range publishers {
+			publishers[i].quitCh <- true
+		}
+		// wait for all publishers to shutdown
+		finishedPublishersWg.Wait()
 
-							// sum errors from all publishers
-							totalErrors := 0
-							for _, publisher := range publishers {
-								totalErrors += publisher.publishErrors
-							}
-							// sum total messages sent from all publishers
-							totalMessages := 0
-							for _, publisher := range publishers {
-								totalMessages += publisher.publishCounter
-							}
-							errorRate := 100 * float64(totalErrors) / float64(totalMessages)
+		// sum errors from all publishers
+		totalErrors := 0
+		for _, publisher := range publishers {
+			totalErrors += publisher.publishErrors
+		}
+		// sum total messages sent from all publishers
+		totalMessages := 0
+		for _, publisher := range publishers {
+			totalMessages += publisher.publishCounter
+		}
+		errorRate := 100 * float64(totalErrors) / float64(totalMessages)
 
-							// report error rate
-							b.ReportMetric(errorRate, "%error")
+		// report error rate
+		b.ReportMetric(errorRate, "%error")
+	}
 
+	// benchmark cases matrix
+	for _, messageSize := range messageSizeCases {
+		b.Run(
+			fmt.Sprintf("msgSz=%db", messageSize),
+			func(b *testing.B) {
+				for _, numPubs := range numPubsCases {
+					b.Run(
+						fmt.Sprintf("pubs=%d", numPubs),
+						func(b *testing.B) {
+							// run workload
+							fanInWorkload(b, numPubs, messageSize)
```

Comment:
There's 2 changes in 1 commit here which make this hard to review:
 - Move to a benchmark function to simplify the "main"
 - Adjust benchmark logic

I would say, forget about 1 for now, focus on 2.

Later and/or in a separate commit, take care of 1 for a all tests in this file.

(also valid: take care of 1 for the whole file, verify nothing changes just style, then move on to make logic changes)

I have not reviewed this file in detail because of the above, but i suspect a number of comments i made on the JS bench below also apply here. Especially setup and wrap up logic.